pkg/placement/placement-resolution.go

```diff
@@ -0,0 +1,368 @@
+/*
+Copyright 2023 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package placement
+
+import (
+	"fmt"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/kubestellar/kubestellar/api/edge/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+// placementResolution stores the selected objects and destinations for a single placement.
+// The mutex should be read locked before reading, and write locked before writing to any field.
+type placementResolution struct {
+	sync.RWMutex
+
+	// map key is GVK/namespace/name as outputted by util.key.GvkNamespacedNameKey()
+	objectIdentifierToKey map[string]*util.Key
+	destinations          sets.Set[string]
+
+	// workloadGeneration is a local counter that is changed whenever a mapped
+	// object undergoes any update. This counter sets that of the
+	// placement-decision's workload whenever syncing.
+	workloadGeneration int64
+}
+
+// noteObject adds/deletes an object to/from the resolution.
+// The return bool indicates whether the placement resolution was changed.
+// This function is thread-safe.
```

Comment:
Doesn't this remark belong in the PlacementResolver interface too?

Reply:
Assuming you only refer to "This function is thread-safe", I've added it to the whole interface (all functions are thread-safe).